### ISSUE TITLE
`Compliance::GAME` codepath in `compute_x` does not use passed `salt` parameter

### ISSUE THREAD

**Issue by username223** [EmberEmu/Ember#26]
I'm in the process of writing my own emulator and have been using Ember in order to verify operations.
As part of this I've used your SRP6 tests/functions in order to track down issues.

I noticed that the `compute_x` method [here](https://github.com/EmberEmu/Ember/blob/d2fae0d9062143368faf4ece322f85a12e6a0820/src/libs/srp6/src/Util.cpp#L97) does not use the passed `salt` paramter, despite RFC2945 defining x as including the salt. Is this an oversight or does the game version not use the salt for x?

```cpp
Botan::BigInt compute_x(const std::string& identifier, const std::string& password,
                        const Botan::BigInt& salt, Compliance mode) {
	//RFC2945 defines x = H(s | H ( I | ":" | p) )
	auto hasher = Botan::HashFunction::create_or_throw("SHA-1");
	std::array<std::uint8_t, SHA1_LEN> hash;
	BOOST_ASSERT_MSG(hash.size() == hasher->output_length(), "Bad hash length");
	hasher->update(identifier);
verifying 
	hasher->update(":");
	hasher->update(password);
	hasher->final(hash.data());

	if(mode == Compliance::RFC5054) {
		hasher->update(Botan::BigInt::encode(salt));
	} else {
		const auto& salt_enc = detail::encode_flip(salt);
		hasher->update(salt_enc.data(), salt_enc.size());
	}

	hasher->update(hash.data(), hash.size());

	if(mode == Compliance::RFC5054) {
		hasher->final(hash.data());
		return Botan::BigInt::decode(hash.data(), hash.size());
	} else {
		return detail::decode_flip(hash);
	}
}
```

eliminating branches that aren't taken:

```cpp
Botan::BigInt compute_x(const std::string& identifier, const std::string& password,
                        const Botan::BigInt& salt, Compliance mode) {
	//RFC2945 defines x = H(s | H ( I | ":" | p) )
	auto hasher = Botan::HashFunction::create_or_throw("SHA-1");
	std::array<std::uint8_t, SHA1_LEN> hash;
	BOOST_ASSERT_MSG(hash.size() == hasher->output_length(), "Bad hash length");
	hasher->update(identifier);
	hasher->update(":");
	hasher->update(password);
	hasher->final(hash.data());
	// hash last modified above

	const auto& salt_enc = detail::encode_flip(salt);
	hasher->update(salt_enc.data(), salt_enc.size());

	hasher->update(hash.data(), hash.size());
	// hasher->final is not called

	// directly return the hash unmodified
	return detail::decode_flip(hash);
}
```

________

Off-topic: Thanks for the excellent resource. This is by far the most well written codebase for WoW private servers that I've found.

A few questions:

1. Is the game capable of using the RFC5054 features?
2. The repo defaults to the `development` branch, is this branch continually push ready or should I use `master` instead?

**Comment by Chaosvex** [comment]
Hi there,

Glad you've found it to be a useful resource for your own emulation work. :)

1. The function initially computes the hash of `SHA1(I | ":" | p)` [here](https://github.com/EmberEmu/Ember/blob/d2fae0d9062143368faf4ece322f85a12e6a0820/src/libs/srp6/src/Util.cpp#L103) and stores the result in `hash`.
`cpp	
hasher->update(identifier);
hasher->update(":");
hasher->update(password);
hasher->final(hash.data());
`

2. It then flips the byte order of the passed salt parameter and stores it in `salt_enc`.
`cpp
const auto& salt_enc = detail::encode_flip(salt);
hasher->update(salt_enc.data(), salt_enc.size());
`

3. Finally, it hashes the result from the first step with the byte-flipped `salt_enc`, so `SHA1(salt_enc | hash)`, which gives us `SHA1(s | SHA1(SHA1(I | ":" | p))`.

Please let me know if that helps or whether I've missed something and good luck. :+1:

**Comment by Chaosvex** [comment]
Hi there,

Glad you've found it to be a useful resource for your own emulation work. :)

I wrote a more elaborate comment until I spotted the comment about `hasher->final()`. You appear to be absolutely correct - it appears my [most recent refactoring](https://github.com/EmberEmu/Ember/commit/d2fae0d9062143368faf4ece322f85a12e6a0820#diff-b1a20743fdabb483ee1be033673fe881192b4044e5bb20c085d5361c77ec6f2d) in this file accidentally moved the call into a branch that I added. Any commit before this should be fine.

It wasn't caught by the unit tests since `compute_x` either isn't used (game auth is tested with static 'replayed' values), was still correct for RFC2945 or used the erroneous `compute_x` in both the client/server calculations. I'll add an regression test to catch this specific case.

Development is the primary branch and should always be in a state that compiles but I often use other long-term branches for major work and then merge to development when ready.

Thanks very much for having a keen eye and reporting this! :+1:

**Comment by Chaosvex** [comment]
Hi there,

Glad you've found it to be a useful resource for your own emulation work. :)

I wrote a more elaborate comment until I spotted the `// hasher->final()`. You are absolutely correct - it appears my [most recent refactoring](https://github.com/EmberEmu/Ember/commit/d2fae0d9062143368faf4ece322f85a12e6a0820#diff-b1a20743fdabb483ee1be033673fe881192b4044e5bb20c085d5361c77ec6f2d) in this file accidentally moved the call into a branch that I added. Any commit before this should be fine.

It wasn't caught by the unit tests since `compute_x` either isn't used (game auth is tested with static 'replayed' values), was still correct for RFC2945 or used the erroneous `compute_x` in both the client/server calculations. I'll add an regression test to catch this specific case.

`development` is the primary branch and should always be in a state that compiles but I often use other long-term branches for major work and then merge to development when ready.

Thanks very much for having a keen eye and reporting this. Best of luck with your own emulation endeavors! :+1:
